README
+----------------------------------------------------------------------+
| mod_auth_ibmdb2                                               README |
+----------------------------------------------------------------------+
| Author: Helmut K. C. Tessarek                                        |
+----------------------------------------------------------------------+
| Website: http://tessus.github.io/mod_auth_ibmdb2                     |
+----------------------------------------------------------------------+

1) Description of the module
2) File descriptions

+----------------------------------------------------------------------+
| 1. Description of the module                                         |
+----------------------------------------------------------------------+

mod_auth_ibmdb2 is an Apache authentication module using IBM DB2 as the backend database for
storing user and group information. The module is designed for Apache 2.0.x and supports
several encryption methods.

Here is a list of the new directives that come with the module:


AuthIBMDB2Database              database name (no default)

AuthIBMDB2Hostname              database server hostname for uncataloged databases (no default)

AuthIBMDB2Portnumber            database instance port (default: 50000)

AuthIBMDB2User                  user for connecting to the DB2 database (no default)

AuthIBMDB2Password              password for connecting to the DB2 database (no default)

AuthIBMDB2UserTable             name of the user table (no default)

AuthIBMDB2GroupTable            name of the group table (no default)

AuthIBMDB2NameField             name of the user field within the table (default: username)

AuthIBMDB2GroupField            name of the group field within the table (default: groupname)

AuthIBMDB2PasswordField         name of the password field within the table (default: password)

AuthIBMDB2CryptedPasswords      passwords are stored encrypted (default: yes)

AuthIBMDB2KeepAlive             connection kept open across requests (default: yes)

AuthIBMDB2Authoritative         lookup is authoritative (default: yes)

AuthIBMDB2NoPasswd              just check, if user is in usertable (default: no)

AuthIBMDB2UserCondition         restrict result set (no default)

AuthIBMDB2GroupCondition        restrict result set (no default)

AuthIBMDB2UserProc              stored procedure for user authentication (no default)

AuthIBMDB2GroupProc             stored procedure for group authentication (no default)

AuthIBMDB2Caching               user credentials are cached (default: off)

AuthIBMDB2GroupCaching          group information is cached (default: off)

AuthIBMDB2CacheFile             path to cache file (default: /tmp/auth_cred_cache)

AuthIBMDB2CacheLifetime         cache lifetime in seconds (default: 300)



Example how to use the module in the httpd.conf:

<Directory "/var/www/my_test_dir">
    AuthName                     "DB2 Authentication"
    AuthType                     Basic

    AuthIBMDB2User               db2inst1
    AuthIBMDB2Password           db2inst1
    AuthIBMDB2Database           auth
    AuthIBMDB2UserTable          users
    AuthIBMDB2NameField          username
    AuthIBMDB2PasswordField      passwd

    AuthIBMDB2CryptedPasswords   On
    AuthIBMDB2KeepAlive          On
    AuthIBMDB2Authoritative      On
    AuthIBMDB2NoPasswd           Off

    AuthIBMDB2GroupTable         groups
    AuthIBMDB2GroupField         groupname

    #require                     valid-user
    require group                admin
    AllowOverride                None
</Directory>

or

<Directory "/var/www/my_test_dir">
    AuthName                     "DB2 Authentication"
    AuthType                     Basic

    AuthIBMDB2User               db2inst1
    AuthIBMDB2Password           db2inst1
    AuthIBMDB2Database           auth
    AuthIBMDB2UserProc           user_sp
    AuthIBMDB2GroupProc          group_sp

    AuthIBMDB2Caching            On
    AuthIBMDB2GroupCaching       On

    require group                admin
    AllowOverride                None
</Directory>

where the stored procedures must have the following parameter format:

CREATE PROCEDURE user_sp  ( IN VARCHAR, OUT VARCHAR )
CREATE PROCEDURE group_sp ( IN VARCHAR )

The stored procedure for user authentication must return exactly one value -
the password. If AuthIBMDB2NoPasswd is set to yes, then the username has
to be returned instead of the password.

The stored procedure for group authentication must return an open cursor
to the resultset.

Two examples for stored procedures are given in the documentation.

+----------------------------------------------------------------------+
| 2. File descriptions                                                 |
+----------------------------------------------------------------------+

mod_auth_ibmdb2.c       the Apache module
mod_auth_ibmdb2.h       the Apache module's header file
caching.h               include file for caching mechanism
makemod                 bash script to compile the module
makemod.bat             script to compile the module under win32
README                  this file
INSTALL                 compile and install instructions
ChangeLog               change log
man/maninstall          bash script to install the manpages


CHANGES
mod_auth_ibmdb2 0.8.4

 - free handle in case of bug 1597078

mod_auth_ibmdb2 0.8.3

 - fixed bug 1597078

mod_auth_ibmdb2 0.8.2

 - added additional debugging code

mod_auth_ibmdb2 0.8.1

 - changed module definition information for compiling the module statically against Apache 1.x
 - added information in the INSTALL file for how to compile the module statically against Apache 1.x
 - added a directory for 2 additional files that are needed when compiling the module statically against Apache 1.x
 - added default value for AuthIBMDB2GroupField

mod_auth_ibmdb2 0.7.1

 - added manpages and a script for installing them
 - implemented caching for group information
 - added 1 new directive for group caching
 - changed the makemod script to use it on AIX aswell

mod_auth_ibmdb2 0.6

 - implemented caching idea contributed by Eddie Anzalone
 - added 3 new directives for caching
 
mod_auth_ibmdb2 0.5.6

 - got rid of the windows stuff again
 - fixed some code due to some compilation issues on some linux systems
 
mod_auth_ibmdb2 0.5.4

 - added a script to compile the module on win32
 - mod_auth_ibmdb2-0.5.4-win32.zip includes the compiled module for win32.
   It was compiled for Apache 2.0.49.
   
mod_auth_ibmdb2 0.5.3

 - eliminated compilation warnings for redefinition of 'TRUE'
 - added additional error checking for user/group condition statements

mod_auth_ibmdb2 0.5.2

 - fixed compilation issue (under Apache 1.x) described in 
   support request [931118]
 
mod_auth_ibmdb2 0.5.1

 - added additional error checking for debugging (access privileges)

mod_auth_ibmdb2 0.5

 - added additional error checking for debugging (table, columns)

mod_auth_ibmdb2 0.4

 - initial release
 
http://mod-auth-ibmdb2.sourceforge.net

$Id: CHANGES,v 1.6.2.1 2007/03/23 21:16:52 tessus Exp $




mod_auth_ibmdb2.h
/*
  +----------------------------------------------------------------------+
  | mod_auth_ibmdb2: structures, defines, globals                        |
  +----------------------------------------------------------------------+
  | Copyright (c) 2006-2015 Helmut K. C. Tessarek                        |
  +----------------------------------------------------------------------+
  | Licensed under the Apache License, Version 2.0 (the "License"); you  |
  | may not use this file except in compliance with the License. You may |
  | obtain a copy of the License at                                      |
  | http://www.apache.org/licenses/LICENSE-2.0                           |
  |                                                                      |
  | Unless required by applicable law or agreed to in writing, software  |
  | distributed under the License is distributed on an "AS IS" BASIS,    |
  | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or      |
  | implied. See the License for the specific language governing         |
  | permissions and limitations under the License.                       |
  +----------------------------------------------------------------------+
  | Author: Helmut K. C. Tessarek                                        |
  +----------------------------------------------------------------------+
  | Website: http://tessus.github.io/mod_auth_ibmdb2                     |
  +----------------------------------------------------------------------+
*/

#ifndef MOD_AUTH_IBMDB2_H
#define MOD_AUTH_IBMDB2_H

#ifndef FALSE								// FALSE
#define FALSE 0
#endif
#ifndef TRUE								// TRUE
#define TRUE (!FALSE)
#endif

#define MAX_IBMDB2_UID_LENGTH   18
#define MAX_IBMDB2_PWD_LENGTH   30
#define MAX_UID_LENGTH          64
#define MAX_PWD_LENGTH         128
#define MAX_GRP_LENGTH         128
#define MAX_DSN_LENGTH         512

#define LOG_DBG( msg ) ap_log_rerror( APLOG_MARK, APLOG_NOERRNO|APLOG_DEBUG, 0, r, "%s", msg )
#define LOG_DBGS( msg ) ap_log_error( APLOG_MARK, APLOG_NOERRNO|APLOG_DEBUG, 0, s, "%s", msg )
#define LOG_ERROR( msg ) ap_log_rerror( APLOG_MARK, APLOG_NOERRNO|APLOG_ERR, 0, r, "%s", msg )
#define MAXERRLEN				SQL_MAX_MESSAGE_LENGTH + SQL_SQLSTATE_SIZE + 255


//	structure to hold the configuration details for the request
typedef struct  {
	char *ibmdb2user;						// user ID to connect to db server
	char *ibmdb2passwd;						// password to connect to db server
	char *ibmdb2DB;							// Database name
	char *ibmdb2host;						// Database hostname (for uncataloged databases)
	int  ibmdb2port;						// Instance port number (for uncataloged databases)
	char *ibmdb2pwtable;					// user password table
	char *ibmdb2grptable;					// user group table
	char *ibmdb2NameField;					// field in password/grp table with username
	char *ibmdb2PasswordField;				// field in password table with password
	char *ibmdb2GroupField;					// field in group table with group name
	int  ibmdb2Crypted;						// are passwords encrypted?
	int  ibmdb2KeepAlive;					// keep connection persistent?
	int  ibmdb2Authoritative;				// are we authoritative?
	int  ibmdb2NoPasswd;					// do we ignore password?
	char *ibmdb2UserCondition; 				// Condition to add to the user where-clause in select query
	char *ibmdb2GroupCondition;				// Condition to add to the group where-clause in select query
	char *ibmdb2UserProc;					// stored procedure for user auth
	char *ibmdb2GroupProc;					// stored procedure for group auth
	int  ibmdb2caching;						// are user credentials cached?
	int  ibmdb2grpcaching;					// is group information cached?
	char *ibmdb2cachefile;					// path to cache file
	char *ibmdb2cachelifetime;				// cache lifetime in seconds
} ibmdb2_auth_config_rec;

//	structure to hold the sqlca variables
typedef struct
{
	char msg[SQL_MAX_MESSAGE_LENGTH + 1];
	char state[SQL_SQLSTATE_SIZE + 1];
	int code;
} sqlerr_t;


static SQLHANDLE   henv;					// environment handle
static SQLHANDLE   hdbc;					// db connection handle

#endif

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


caching.h
/*
  +----------------------------------------------------------------------+
  | caching: functions for caching mechanism                             |
  +----------------------------------------------------------------------+
  | Copyright (c) 2006-2015 Helmut K. C. Tessarek                        |
  +----------------------------------------------------------------------+
  | Licensed under the Apache License, Version 2.0 (the "License"); you  |
  | may not use this file except in compliance with the License. You may |
  | obtain a copy of the License at                                      |
  | http://www.apache.org/licenses/LICENSE-2.0                           |
  |                                                                      |
  | Unless required by applicable law or agreed to in writing, software  |
  | distributed under the License is distributed on an "AS IS" BASIS,    |
  | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or      |
  | implied. See the License for the specific language governing         |
  | permissions and limitations under the License.                       |
  +----------------------------------------------------------------------+
  | Author: Helmut K. C. Tessarek                                        |
  +----------------------------------------------------------------------+
  | Website: http://tessus.github.io/mod_auth_ibmdb2                     |
  +----------------------------------------------------------------------+
*/

#ifndef CACHING_H
#define CACHING_H

#include "apr_general.h"
#include "apr_dbm.h"
#include "apr_time.h"

typedef struct {
        char   password[MAX_PWD_LENGTH];
        apr_time_t timestamp;
} cached_password_timestamp;

typedef struct {
        int    numgrps;
        apr_time_t timestamp;
} cached_group_timestamp;

static int write_cache( request_rec *r, const char *user, const char *password, ibmdb2_auth_config_rec *m );
static char *read_cache( request_rec *r, const char *user, ibmdb2_auth_config_rec *m );
static int write_group_cache( request_rec *r, const char *user, const char **grplist, ibmdb2_auth_config_rec *m );
static char **read_group_cache( request_rec *r, const char *user, ibmdb2_auth_config_rec *m );

/*
	function to store the user credentials in the local cache, so that subsequent
	http requests can validate the user directly from local cache without the need
	to query the backend db2 database.
*/

/* {{{ static int write_cache( request_rec *r, const char *user, const char *password, ibmdb2_auth_config_rec *m )
*/
static int write_cache( request_rec *r, const char *user, const char *password, ibmdb2_auth_config_rec *m )
{
	apr_status_t rc;
	apr_pool_t *pool;
	apr_dbm_t *db;
	char errmsg[MAXERRLEN];
	int rc_write_cache = 0;

	char *my_user = (char *)user;
	apr_datum_t datum_user = { my_user, (strlen( my_user )+1) };

	cached_password_timestamp cpt;

	apr_datum_t datum_value;

	char *my_password = (char *)password;
	strcpy(cpt.password, my_password);

	if( !(cpt.timestamp = apr_time_now()) )
	{
		errmsg[0] = '\0';
		sprintf( errmsg, "unable to determine current time (write cache)");
		LOG_ERROR( errmsg );
		return( 1 );
	}

	datum_value.dptr = (void *)&cpt;
	datum_value.dsize = sizeof(cpt);
	
	apr_pool_create( &pool, NULL );

	rc = apr_dbm_open(&db, m->ibmdb2cachefile, APR_DBM_RWCREATE, APR_UREAD | APR_UWRITE, pool);

	if ( rc == APR_SUCCESS )
	{
		errmsg[0] = '\0';
		sprintf( errmsg, "storing user [%s] and pass [%s] in cache", my_user, my_password);
		LOG_DBG( errmsg );

		if( (rc = apr_dbm_store(db, datum_user, datum_value)) != APR_SUCCESS )
		{
			errmsg[0] = '\0';
			sprintf( errmsg, "unable to store user [%s] in cache", my_user);
			LOG_DBG( errmsg );
			errmsg[0] = '\0';
			apr_strerror( rc, errmsg, sizeof(errmsg) );
			LOG_DBG( errmsg );
			
			rc_write_cache = 1;
		}

		apr_dbm_close( db );
	}
	else
	{
		errmsg[0] = '\0';
		sprintf( errmsg, "could not open cachefile [%s] for writing", m->ibmdb2cachefile );
		LOG_ERROR( errmsg );
		errmsg[0] = '\0';
		apr_strerror( rc, errmsg, sizeof(errmsg) );
		LOG_DBG( errmsg );
		
		rc_write_cache = 1;
	}
	
	apr_pool_destroy( pool );

	return( rc_write_cache );
}
/* }}} */

/*
	function to check in the local cache to validate user, otherwise
	we need to query the backend db2 database.
*/

/* {{{ static char *read_cache( request_rec *r, const char *user, ibmdb2_auth_config_rec *m )
*/
static char *read_cache( request_rec *r, const char *user, ibmdb2_auth_config_rec *m )
{
	apr_status_t rc;
	apr_pool_t *pool;
	apr_dbm_t *db;
	char errmsg[MAXERRLEN];

	char *my_user = (char *)user;
	apr_datum_t datum_user = { my_user, (strlen( my_user )+1) };

	cached_password_timestamp cpt;
	apr_time_t current_time;

	apr_datum_t datum_value;

	char *pw = NULL;

	int MAXAGE = atoi( m->ibmdb2cachelifetime );

	apr_time_t time_diff;
	
	apr_pool_create( &pool, NULL );

	rc = apr_dbm_open(&db, m->ibmdb2cachefile, APR_DBM_RWCREATE, APR_UREAD | APR_UWRITE, pool);

	if( rc == APR_SUCCESS )
	{
		rc = apr_dbm_fetch( db, datum_user, &datum_value );

		if( rc == APR_SUCCESS )
		{
			if( datum_value.dsize != sizeof(cpt) )
			{
				errmsg[0] = '\0';
				sprintf( errmsg, "we found our user in the cache but with corrupted record: %s \n", my_user);
				LOG_ERROR( errmsg );
				
				apr_dbm_close( db );
				apr_pool_destroy( pool );

				return NULL;
			}
			else
			{
				memcpy((void *)&cpt, datum_value.dptr, datum_value.dsize);

				if( !(current_time = apr_time_now()) )
				{
					errmsg[0] = '\0';
					sprintf( errmsg, "unable to determine current time (read cache)");
					LOG_ERROR( errmsg );
					
					apr_dbm_close( db );
					apr_pool_destroy( pool );

					return NULL;
				}

				time_diff = current_time - cpt.timestamp;

				if( apr_time_sec(time_diff) > MAXAGE )
				{
					errmsg[0] = '\0';
					sprintf( errmsg, "cached password for user [%s] is toooo old", my_user);
					LOG_DBG( errmsg );
					
					apr_dbm_close( db );
					apr_pool_destroy( pool );

					return NULL;
				}

				pw = (char *)PSTRDUP(r->pool, cpt.password);

				/* Congratulations, we have a fresh cached entry */
				errmsg[0] = '\0';
				sprintf( errmsg, "user [%s] - [%s] found in cache", my_user, pw);
				LOG_DBG( errmsg );
				
				apr_dbm_close( db );
				apr_pool_destroy( pool );

				return pw;
			}
		}
		else
		{
			/* Did not find user in the cache */
			errmsg[0] = '\0';
			sprintf( errmsg, "user [%s] not found in cache", my_user);
			LOG_DBG( errmsg );
			
			apr_dbm_close( db );
			apr_pool_destroy( pool );
			
			return NULL;
		}
	}
	else
	{
		errmsg[0] = '\0';
		sprintf( errmsg, "could not open cachefile [%s] for reading", m->ibmdb2cachefile );
		LOG_ERROR( errmsg );
		errmsg[0] = '\0';
		apr_strerror( rc, errmsg, sizeof(errmsg) );
		LOG_DBG( errmsg );

		apr_pool_destroy( pool );
		return NULL;
	}
}
/* }}} */

/*
	function to store the group information in the local cache, so that subsequent
	http requests can validate the groups directly from local cache without the need
	to query the backend db2 database.
*/

/* {{{ static int write_group_cache( request_rec *r, const char *user, const char **grplist, ibmdb2_auth_config_rec *m )
*/
static int write_group_cache( request_rec *r, const char *user, const char **grplist, ibmdb2_auth_config_rec *m )
{
	apr_status_t rc;
	apr_pool_t *pool;
	apr_dbm_t *db;
	char errmsg[MAXERRLEN];
	int rc_write_group_cache = 0;

	char ibmdb2grpcachefile[512];
	char username[MAX_UID_LENGTH+4];
	char groupname[MAX_GRP_LENGTH];

	int i = 0;

	char *my_user = (char *)user;
	apr_datum_t datum_user = { my_user, (strlen( my_user )+1) };

	cached_group_timestamp cgt;

	apr_datum_t datum_value;
	apr_datum_t key_data;
	apr_datum_t data_data;
	
	sprintf( ibmdb2grpcachefile, "%s.grp", m->ibmdb2cachefile );

	if( !(cgt.timestamp = apr_time_now()) )
	{
		errmsg[0] = '\0';
		sprintf( errmsg, "unable to determine current time (write group cache)");
		LOG_ERROR( errmsg );
		return( 1 );
	}

	while( grplist[i] )
	{
		++i;
	}

	cgt.numgrps = i;

	datum_value.dptr = (void *)&cgt;
	datum_value.dsize = sizeof(cgt);

	apr_pool_create( &pool, NULL );
	
	rc = apr_dbm_open(&db, ibmdb2grpcachefile, APR_DBM_RWCREATE, APR_UREAD | APR_UWRITE, pool);

	if( rc == APR_SUCCESS )
	{
		if( (rc = apr_dbm_store(db, datum_user, datum_value)) != APR_SUCCESS )
		{
			errmsg[0] = '\0';
			sprintf( errmsg, "unable to store group info for user [%s] in cache", my_user);
			LOG_DBG( errmsg );
			errmsg[0] = '\0';
			apr_strerror( rc, errmsg, sizeof(errmsg) );
			LOG_DBG( errmsg );

			apr_dbm_close( db );
			apr_pool_destroy( pool );
	        return( 1 );
		}

		i = 0;

		while( grplist[i] )
		{
			key_data.dptr = NULL;
            data_data.dptr = NULL;

			username[0] = '\0';
			sprintf( username, "%s_%d", my_user, i );

			groupname[0] = '\0';
			strcpy( groupname, grplist[i] );

			key_data.dptr = username;
			key_data.dsize = strlen(username) + 1;

			data_data.dptr = groupname;
			data_data.dsize = strlen(groupname) + 1;

			if( (rc = apr_dbm_store(db, key_data, data_data)) != APR_SUCCESS )
			{
				errmsg[0] = '\0';
				sprintf( errmsg, "unable to store group [%s] for user [%s] in cache", grplist[i], my_user );
				LOG_DBG( errmsg );
				errmsg[0] = '\0';
				apr_strerror( rc, errmsg, sizeof(errmsg) );
				LOG_DBG( errmsg );

				apr_dbm_close( db );
				apr_pool_destroy( pool );
				return( 1 );
			}
			else
			{
				errmsg[0] = '\0';
				sprintf( errmsg, "storing user [%s] and group [%s] in cache", my_user, grplist[i] );
				LOG_DBG( errmsg );
			}

			++i;
		}

		apr_dbm_close( db );
	}
	else
	{
		errmsg[0] = '\0';
		sprintf( errmsg, "could not open group cachefile [%s] for writing", ibmdb2grpcachefile );
		LOG_ERROR( errmsg );
		errmsg[0] = '\0';
		apr_strerror( rc, errmsg, sizeof(errmsg) );
		LOG_DBG( errmsg );

		rc_write_group_cache = 1;
	}
	
	apr_pool_destroy( pool );
	
	return( rc_write_group_cache );
}
/* }}} */

/*
	function to check in the local cache to check if user is in a group, otherwise
	we need to query the backend db2 database.
*/

/* {{{ static char **read_group_cache( request_rec *r, const char *user, ibmdb2_auth_config_rec *m )
*/
static char **read_group_cache( request_rec *r, const char *user, ibmdb2_auth_config_rec *m )
{
	apr_status_t rc;
	apr_pool_t *pool;
	apr_dbm_t *db;
	char errmsg[MAXERRLEN];

	char ibmdb2grpcachefile[512];
	char username[MAX_UID_LENGTH+4];

	int i = 0;

	int numgrps;

	char *my_user = (char *)user;
	apr_datum_t datum_user = { my_user, (strlen( my_user )+1) };

	cached_group_timestamp cgt;
	apr_time_t current_time;

	apr_datum_t datum_value;
	apr_datum_t key_data;
	apr_datum_t return_data;

	char **list = NULL;

	int MAXAGE = atoi( m->ibmdb2cachelifetime );

	apr_time_t time_diff;

	sprintf( ibmdb2grpcachefile, "%s.grp", m->ibmdb2cachefile );
	
	apr_pool_create( &pool, NULL );

	rc = apr_dbm_open(&db, ibmdb2grpcachefile, APR_DBM_RWCREATE, APR_UREAD | APR_UWRITE, pool);

	if( rc == APR_SUCCESS )
	{
		rc = apr_dbm_fetch( db, datum_user, &datum_value );

		if( rc == APR_SUCCESS )
		{
			if( datum_value.dsize != sizeof(cgt) )
			{
				errmsg[0] = '\0';
				sprintf( errmsg, "we found our user in the cache but with corrupted record: %s \n", my_user);
				LOG_ERROR( errmsg );
				
				apr_dbm_close( db );
				apr_pool_destroy( pool );

				return NULL;
			}
			else
			{
				memcpy((void *)&cgt, datum_value.dptr, datum_value.dsize);

				if( !(current_time = apr_time_now()) )
				{
					errmsg[0] = '\0';
					sprintf( errmsg, "unable to determine current time (read group cache)");
					LOG_ERROR( errmsg );
					
					apr_dbm_close( db );
					apr_pool_destroy( pool );

					return NULL;
				}

				time_diff = current_time - cgt.timestamp;

				i = 0;

				numgrps = cgt.numgrps;

				if( apr_time_sec(time_diff) > MAXAGE )
				{
					errmsg[0] = '\0';
					sprintf( errmsg, "cached group information for user [%s] toooo old", my_user);
					LOG_DBG( errmsg );
					
					apr_dbm_close( db );
					apr_pool_destroy( pool );

					return NULL;
				}

				/* Build the list to return */

				list = (char **) malloc(sizeof(char *) * (numgrps+1));

				for( i = 0; i < numgrps; i++ )
				{
					key_data.dptr = NULL;
					return_data.dptr = NULL;

					username[0] = '\0';
					sprintf( username, "%s_%d", my_user, i );

					key_data.dptr = username;
					key_data.dsize = strlen(username) + 1;

					rc = apr_dbm_fetch( db, key_data, &return_data );

					if( rc == APR_SUCCESS )
					{
						list[i] = return_data.dptr;
					}
				}

				list[i] = NULL;

				/* Congratulations, we have a fresh cached entry */
				errmsg[0] = '\0';
				sprintf( errmsg, "groups for user [%s] found in cache", my_user);
				LOG_DBG( errmsg );
				
				apr_dbm_close( db );
				apr_pool_destroy( pool );

				return list;
			}
		}
		else
		{
			/* Did not find user in the group cache */
			errmsg[0] = '\0';
			sprintf( errmsg, "groups for user [%s] not found in cache", my_user);
			LOG_DBG( errmsg );
			
			apr_dbm_close( db );
			apr_pool_destroy( pool );

			return NULL;
		}
	}
	else
	{
		errmsg[0] = '\0';
		sprintf( errmsg, "could not open group cachefile [%s] for reading", ibmdb2grpcachefile );
		LOG_ERROR( errmsg );
		errmsg[0] = '\0';
		apr_strerror( rc, errmsg, sizeof(errmsg) );
		LOG_DBG( errmsg );

		apr_pool_destroy( pool );
		return NULL;
	}
}
/* }}} */

#endif

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */
